src/dag-node-factory.js

```diff
@@ -0,0 +1,193 @@
+'use strict'
+
+const stable = require('stable')
+const sort = stable.inplace
+// const mh = require('multihashes')
+const protobuf = require('protocol-buffers')
+const proto = protobuf(require('./dag.proto'))
+const util = require('./util')
+const DAGNode = require('./dag-node')
+const DAGLink = require('./dag-link')
+
+class DAGNodeFactory {
+  static create (data, dagLinks, hashAlg, callback) {
+    if (typeof data === 'function') {
+      // empty obj
+      callback = data
+      data = undefined
+    }
+    if (typeof dagLinks === 'function') {
+      // empty obj
+      callback = dagLinks
+      dagLinks = []
+    }
+    if (typeof hashAlg === 'function') {
+      // empty obj
+      callback = hashAlg
+      hashAlg = undefined
+    }
+
+    if (!hashAlg) {
+      hashAlg = 'sha2-256'
+    }
+
+    const links = dagLinks.map((l) => {
+      if (!l.constructor && l.constructor.name !== 'DAGLink') {
+        return l
+      }
+
+      // haadcode: are the .name vs .Name for backwards compatibility?
+      const link = new DAGLink(l.name || l.Name,
+                               l.size || l.Size,
+                               l.hash || l.Hash || l.multihash)
+      return link
+    })
+
+    // Sort the links (in-place)
+    sort(links, util.linkSort)
+
+    DAGNodeFactory._serialize({
+      data: data,
+      links: links
+    }, (err, serialized) => {
+      if (err) {
+        callback(err)
+      }
+      util.hash(hashAlg, serialized, (err, multihash) => {
+        if (err) {
+          callback(err)
+        }
+        const dagNode = new DAGNode(data, links, serialized, multihash)
+        callback(null, dagNode)
+      })
+    })
+  }
+
+  static addLink (dagNode, nameOrLink, nodeOrMultihash, callback) {
+    const links = DAGNodeFactory._cloneLinks(dagNode)
+    const data = DAGNodeFactory._cloneData(dagNode)
+    let newLink = null
+
+    if ((nameOrLink.constructor &&
+         nameOrLink.constructor.name === 'DAGLink')) {
+      // It's a link
+      newLink = nameOrLink
+    } else if (typeof nameOrLink === 'string') {
+      // It's a name
+      if ((nodeOrMultihash.constructor &&
+         nodeOrMultihash.constructor.name === 'DAGNode')) {
+        // It's a node
+        // haadcode: not sure what ^ means, so the line below might not be correct
+        newLink = DAGNodeFactory.toDAGLink(nodeOrMultihash)
+      } else {
+        // It's a multihash
+        // haadcode: not sure what ^ means, so the line below might not be correct
+        newLink = new DAGLink(null, dagNode.size, nodeOrMultihash)
+      }
+    }
+
+    if (newLink) {
+      links.push(newLink)
+      sort(links, util.linkSort)
+    } else {
+      throw new Error('Link given as the argument is invalid')
+    }
+
+    DAGNodeFactory.create(data, links, callback)
+  }
```

Comment:
Not sure if we want this methods in a thing that is called factory.

Not only forces people to have to require the module if they want to add a link:

```
const node = <some dagnode I got from ipfs.object.get
// now I want to add a link
const dagNodeFactory = require('ipld-dag-pb').dagNodeFactory
dagNodeFactory.addLink(node...)
```

while it could just be (as it was):

```
const newNode = node.addLink(link)
```